java/rocksjni/merge_operator_jnicallback.cc

```diff
@@ -0,0 +1,300 @@
+//  Copyright (c) 2017-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "rocksjni/merge_operator_jnicallback.h"
+#include "rocksjni/portal.h"
+
+namespace rocksdb {
+
+  MergeOperatorJniCallback::MergeOperatorJniCallback(JNIEnv* env,
+                                                     jobject jMergeOperator,
+                                                     jboolean _allowSingleOperand,
+                                                     jboolean _allowShouldMerge,
+                                                     jboolean _allowPartialMultiMerge)
+      : JniCallback(env, jMergeOperator)
+  {
+    allowSingleOperand = _allowSingleOperand;
+    allowShouldMerge = _allowShouldMerge;
+    allowPartialMultiMerge = _allowPartialMultiMerge;
+
+    // Note: The name of a MergeOperator will not change during it's lifetime,
+    // so we cache it in a global var
+    jmethodID jNameMethodId = MergeOperatorJni::getNameMethodId(env);
+    if(jNameMethodId == nullptr) {
+      // exception thrown: NoSuchMethodException or OutOfMemoryError
+      return;
+    }
+    jstring jsName = (jstring)env->CallObjectMethod(m_jcallback_obj, jNameMethodId);
+    if (env->ExceptionCheck()) {
+      // exception thrown
+      return;
+    }
+    jboolean has_exception = JNI_FALSE;
+    m_name = JniUtil::copyString(env, jsName,
+                                 &has_exception);  // also releases jsName
+    if (has_exception == JNI_TRUE) {
+      // exception thrown
+      return;
+    }
+
+    fullMergeMethodId = MergeOperatorJni::getFullMergeMethodId(env);
+
+    partialMultiMergeMethodId = MergeOperatorJni::getPartialMultiMergeMethodId(env);
+
+    partialMergeMethodId = MergeOperatorJni::getPartialMergeMethodId(env);
+
+    shouldMergeMethodId = MergeOperatorJni::getShouldMergeMethodId(env);
+
```

Comment:
The above methodId lookups need nullptr checks.

Reply:
done